File: src/javatruco/Carta.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package javatruco;

/**
 *
 * @author kaique
 */
public class Carta {
    public static final int 
            C4 = 0, C5 = 1, C6 = 2, C7 = 3,
            Q = 4, J = 5, K = 6, A = 7, C2 = 8, C3 = 9;
    
    public static final int OURO = 0;
    public static final int ESPADAS = 1;
    public static final int COPAS = 2;
    public static final int PAUS = 3;
    
    public static final int PERDE = -1;
    public static final int EMPATA = 0;
    public static final int GANHA = 1;
    
    
    protected int valor;
    protected int naipe;
    
    public Carta(int valor, int naipe){
        this.valor = valor;
        this.naipe = naipe;
    }
    
    /**
     * Obtém o "desenho" da carta
     * @return String com valor e naipe da carta
     */
    public String Desenho(){
        String carta = "";
        switch(this.getValor()){
            case C4:
                carta += "4";
                break;
            case C5:
                carta += "5";
                break;
            case C6:
                carta += "6";
                break;
            case C7:
                carta += "7";
                break;
            case Q:
                carta += "Q";
                break;
            case J:
                carta += "J";
                break;
            case K:
                carta += "K";
                break;
            case A:
                carta += "A";
                break;
            case C2:
                carta += "2";
                break;
            case C3:
                carta += "3";
                break;
        }
        switch(getNaipe()){
            case OURO:
                carta += "♦";
                break;
            case COPAS:
                carta += "♥";
                break;
            case ESPADAS:
                carta += "♠";
                break;
            case PAUS:
                carta += "♣";
                break;
        }
        return carta;
    }
    
    /**
     * Verifica se a carta é manilha
     * @return Retorna TRUE se a carta for manilha
     */
    private boolean Manilha(Carta c, Carta vira){
        return c.getValor() == (vira.getValor()+1) || 
                (c.getValor()==C4 && vira.getValor()==C3);
    }
    
    /**
     * Verifica se a carta ganha
     * @return Retorna GANHA, EMPATA ou PERDE, dependendo das condições
     */
    public int Ganha(Carta outra, Carta vira){
        if(Manilha(this, vira)){
            if(Manilha(outra, vira) && this.getNaipe() < outra.getNaipe()){
                return PERDE;
            }
            return GANHA;
        }
        else{
            if(Manilha(outra, vira) || this.getValor() < outra.getValor()){
                return PERDE;
            }
            if(this.getValor() > outra.getValor())
                return GANHA;
        }
            
        return EMPATA;
    }

    /**
     * @return o valor
     */
    public int getValor() {
        return valor;
    }

    /**
     * @param valor o valor para definir
     */
    public void setValor(int valor) {
        this.valor = valor;
    }

    /**
     * @return o naipe
     */
    public int getNaipe() {
        return naipe;
    }

    /**
     * @param naipe o naipe para definir
     */
    public void setNaipe(int naipe) {
        this.naipe = naipe;
    }
}

File: src/ctruco/Carta.c
/**************
 * Constantes *
 **************/
#define _C4 0
#define _C5 1
#define _C6 2
#define _C7 3
#define _CQ 4
#define _CJ 5
#define _CK 6
#define _CA 7
#define _C2 8
#define _C3 9
#define _CNONE -1

#define _OURO 0
#define _ESPADAS 1
#define _COPAS 2
#define _PAUS 3
#define _NONAIPE -1

#define _PERDE -1
#define _EMPATA 0
#define _GANHA 1

/*************
 * Atributos *
 *************/
var(public, int, valor);
var(public, int, naipe);

/**********
 * Classe *
 **********/
class(Carta, {&valor, &naipe});

/*******************************************************
 * Construtor - 									   *
 * Recebe dois int, contento o valor e naipe da carta. * 
 *******************************************************/
constructor(Carta, 
	arg(int, v);
	arg(int, n);
	get(valor) = (generic)(intptr_t)v;
	get(naipe) = (generic)(intptr_t)n;
	//printf("Carta criada: %d:%d - %s\n", v, n, (char*)call(&this, Desenho));
	);

/***********
 * Funções *
 ***********/
/*
 * Carta:Desenho - 
 * Retorna um generic char* com o Desenho da carta
 */
function(Carta, public, Desenho, 
	char *carta = (char*)malloc(4);
	switch((int)(intptr_t)get(valor)){
		case _C4:
			carta[0] = '4';
			break;
		case _C5:
			carta[0] = '5';
			break;
		case _C6:
			carta[0] = '6';
			break;
		case _C7:
			carta[0] = '7';
			break;
		case _CQ:
			carta[0] = 'Q';
			break;
		case _CJ:
			carta[0] = 'J';
			break;
		case _CK:
			carta[0] = 'K';
			break;
		case _CA:
			carta[0] = 'A';
			break;
		case _C2:
			carta[0] = '2';
			break;
		case _C3:
			carta[0] = '3';
			break;
		case _CNONE:
			carta[0] = ' ';
			break;
		default:
			fprintf(stderr, "Erro: Valor de carta não identificado.\n");
	}
	switch((int)(intptr_t)get(naipe)){
		case _OURO:
			carta[1] = 'o';
			break;
		case _COPAS:
			carta[1] = 'c';
			break;
		case _ESPADAS:
			carta[1] = 'e';
			break;
		case _PAUS:
			carta[1] = 'p';
			break;
		case _NONAIPE:
			carta[1] = ' ';
			break;
		default:
			fprintf(stderr, "Erro: Naipe não identificado.\n");
	}
	carta[2] = '\0';
	carta[3] = '\0';
	return carta;
	);
/*
 * Carta:Desenho - 
 * Recebe dois Objetos, sendo o vira e a carta, e verifica se ela é manilha,
 * retornando um generic int 1 caso verdadeiro, 0 caso falso.
 */
function(Carta, public, Manilha, 
	arg(Object*, c);
	arg(Object*, vira);
	int cVal, vVal;
	//printf("<< Deseinho carta: %s>>\n", (char*)call(c, Desenho));
	//printf("<< Deseinho vira: %s>>\n", (char*)call(vira, Desenho));
	cVal = (int)(intptr_t)call(c, getValor);
	vVal = (int)(intptr_t)call(vira, getValor);
	return (generic)(intptr_t)((cVal == vVal+1) || (cVal == _C4 && vVal == _C3));
	);
/*
 * Carta:Ganha - 
 * Recebe dois Objetos, sendo o vira e a outra carta, e verifica se esta carta
 * a vence, retornando um generic int com _GANHA, _EMPATA ou _PERDE.
 */
function(Carta, public, Ganha,
	arg(Object*, outra);
	arg(Object*, vira);
	//puts("<< Disputando >>");
	if(outra==NULL || (int)(intptr_t)call(outra, getValor)==_CNONE || (int)(intptr_t)call(outra, getValor)==_NONAIPE)
		return (generic)(intptr_t)_GANHA;
	//printf("<< Deseinho carta: %s>>\n", (char*)call(&this, Desenho));
	//printf("<< Deseinho outra: %s>>\n", (char*)call(outra, Desenho));
	//printf("<< Deseinho vira: %s>>\n", (char*)call(vira, Desenho));
	//puts("<< Verificando >>");
	if((int)(intptr_t)call(&this, Manilha, &this, vira)){
		//puts("<< Vendo se perde >>");
		if((int)(intptr_t)call(&this, Manilha, outra, vira) && ((int)(intptr_t)call(&this, getNaipe) < (int)(intptr_t)call(outra, getNaipe))){
			//puts("<< PERDEU >>");
			return (generic)(intptr_t)_PERDE;
		}
		//puts("<< GANHOU >>");
		return (generic)(intptr_t)_GANHA;
	}
	else{
		//puts("<< Não é maniha >>");
		//printf("<< Deseinho outra: %s>>\n", (char*)call(outra, Desenho));
		if((int)(intptr_t)call(&this, Manilha, outra, vira) || ((int)(intptr_t)call(&this, getValor) < (int)(intptr_t)call(outra, getValor)) ){
			//puts("<< PERDEU 2 >>");
			return (generic)(intptr_t)_PERDE;
		}
		//puts("<< GANHOU 2 >>");
		return (generic)(intptr_t)_GANHA;
	}
	//puts("<< Empatou >>");
	return (generic)(intptr_t)_EMPATA;
	);
/*
 * Carta:getValor - 
 * Retorna um generic int com o valor da carta
 */
function(Carta, public, getValor,
	return get(valor);
	);
/*
 * Carta:setValor - 
 * Recebe um int, e define o valor da carta.
 */
function(Carta, public, setValor,
	arg(int, v);
	get(valor) = (generic)(intptr_t)v;
	);
/*
 * Carta:getNaipe - 
 * Retorna um generic int com o Naipe da carta
 */
function(Carta, public, getNaipe,
	return get(naipe);
	);
/*
 * Carta:setNaipe - 
 * Recebe um int, e define o valor da carta.
 */
function(Carta, public, setNaipe,
	arg(int, n);
	get(naipe) = (generic)(intptr_t)n;
	);
/*
 * Carta:Desenho - 
 * Define o valor e naipe da carta como nulos
 */
function(Carta, public, nulificar,
	get(naipe) = (generic)(intptr_t)_NONAIPE;
	get(valor) = (generic)(intptr_t)_CNONE;
	);